control/cli.py

```diff
@@ -156,298 +207,1590 @@ def connect(self, host, port, client_key, client_cert, server_cert):
+            return ret
+        else:
+            assert False
+
+        return ret.status
+
+    ns_args_list = [
+        argument("ns_command", help="namespace sub-command",
+                 choices=["add", "del", "resize", "list", "get_io_stats", "change_load_balancing_group", "set_qos"]),
+        argument("--subsystem", "-n", help="Subsystem NQN", required=True),
+        argument("--rbd-pool", "-p", help="RBD pool name"),
+        argument("--rbd-image", "-i", help="RBD image name"),
+        argument("--block-size", "-s", help="Block size", type=int),
+        argument("--uuid", "-u", help="UUID"),
+        argument("--nsid", help="Namespace ID", type=int),
+        argument("--load-balancing-group", "-l", help="Load balancing group", type=int),
+        argument("--size", help="New size in MiB", type=int),
+        argument("--rw-ios-per-second", help="R/W IOs per second limit, 0 means unlimited", type=int),
+        argument("--rw-megabytes-per-second", help="R/W megabytes per second limit, 0 means unlimited", type=int),
+        argument("--r-megabytes-per-second", help="Read megabytes per second limit, 0 means unlimited", type=int),
+        argument("--w-megabytes-per-second", help="Write megabytes per second limit, 0 means unlimited", type=int),
+    ]
+    @cli.cmd(ns_args_list)
+    def ns(self, args):
+        """Namespace commands"""
+        if args.ns_command == "add":
+            return self.ns_add(args)
+        elif args.ns_command == "del":
+            return self.ns_del(args)
+        elif args.ns_command == "resize":
+            return self.ns_resize(args)
+        elif args.ns_command == "list":
+            return self.ns_list(args)
+        elif args.ns_command == "get_io_stats":
+            return self.ns_get_io_stats(args)
+        elif args.ns_command == "change_load_balancing_group":
+            return self.ns_change_load_balancing_group(args)
+        elif args.ns_command == "set_qos":
+            return self.ns_set_qos(args)
+        assert False
+
+    @cli.cmd(ns_args_list)
+    def namespace(self, args):
+        """Namespace commands"""
+        return self.ns(args)
+
+def main_common(client, args):
+    server_address = args.server_address
+    server_port = args.server_port
+    client_key = args.client_key
+    client_cert = args.client_cert
+    server_cert = args.server_cert
+    client.connect(server_address, server_port, client_key, client_cert, server_cert)
+    call_function = getattr(client, args.func.__name__)
+    rc = call_function(args)
+    return rc
+
+def main_test(args):
+    if not args:
+        return None
+    try:
+        i = args.index("--format")
+        del args[i:i + 2]
+    except Exception:
+        pass
+    args = ["--format", "python"] + args
+    client = GatewayClient()
+    parsed_args = client.cli.parser.parse_args(args)
+    if parsed_args.subcommand is None:
+        return None
+
+    return main_common(client, parsed_args)
```

Comment:
In general we shouldn't put test code in production artifacts. Python, unlike other languages, is flexible enough not to need test code/branching inside production code.

In any case, it looks like it has to do with the output format and that "hack" of `--format=python` to allow by-passing Python objects for testing purposes (is there any other reason?).

I checked and the `args.format ==` expression appears 169 times in the `cli.py` file:
```py
        if args.format == "text" or args.format == "plain":
            if ret.status == 0:
                if args.nsid:
                    ns_id_str = f"{args.nsid}"
                elif args.uuid:
                    ns_id_str = f"with UUID {args.uuid}"
                else:
                    assert False
                out_func(f"Changing load balancing group of namespace {ns_id_str} in {args.subsystem} to {args.load_balancing_group}: Successful")
            else:
                err_func(f"{ret.error_message}")
        elif args.format == "json" or args.format == "yaml":
            ret_str = json_format.MessageToJson(
                        ret,
                        indent=4,
                        including_default_value_fields=True,
                        preserving_proto_field_name=True)
            if args.format == "json":
                out_func(f"{ret_str}")
            elif args.format == "yaml":
                obj = json.loads(ret_str)
                out_func(yaml.dump(obj))
        elif args.format == "python":
            return ret
        else:
            assert False
```

First of all, I'd follow as much as possible [the Unix approach](https://clig.dev/#:~:text=Make%20it%20easy%20to%20see%20the%20current%20state%20of%20the%20system.): whenever a command worked as expected, no output is shown. 

That aside, rather than doing this checking inside every command, you could use a declarative approach and a `Formatter`-like (singleton) class/decorator, where you just return the object, and describe how you wanted it to be formatted:
```py
  @cli.formatter(
    text=lambda x: f"Version is {x}")
  )
  def version(self, args):
    ...
    return ret
```

The `formatter` class would:
* read the `--format` arg directly from `sys.argv` or explicitly at instantiation.
* format the output by running the input callables `text`, `json`, `yaml`, ...
* if not provided, inspect the passed object for magic methods like `__str__`, `__json__`, `__yaml__`, and invoke them,
* otherwise, would implement default methods for serializing to `text`, `json`, `yaml`, etc.

Something like that would reduce the code here and might create a more consistent output format. For the `format=python` case, just mocking the `Formatter` in the test files would be enough.

What do you think, @gbregman ?